### PR TITLE
FEA-83 Create hook for handling adding to favorites action

### PR DESCRIPTION
**Pull request by adrian-konopnicki-CC**

### CODE REVIEW

File: src/components/FavoritesButton/FavoritesButton.tsx [hunk 1]
```diff
@@ -1,14 +1,21 @@
+import { MouseEvent } from 'react';
 import { Heart } from 'react-feather';
 import { Button } from './FavoritesButton.styled';
 import { FavoritesButtonTestIds } from './FavoritesButtonTestIds';
 
 interface Props {
   isFavorite: boolean;
```
Comment (kamil-borecki-cc) [inline src/components/FavoritesButton/FavoritesButton.tsx]: **suggestion** dont use `is` prefix for boolean props

File: src/components/FavoritesButton/FavoritesButton.tsx [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+  onClick: () => void;
 }
 
-const FavoritesButton = ({ isFavorite }: Props) => {
+const FavoritesButton = ({ isFavorite, onClick }: Props) => {
+  const handleClick = (e: MouseEvent<HTMLButtonElement>) => {
```
Comment (kamil-borecki-cc) [inline src/components/FavoritesButton/FavoritesButton.tsx:17]: **suggestion** handle favorites action inside this component.

This one will be used in many places, i think the good idea will be isolate it and handle addToFavorites action within this component

File: src/context/favorites/FavoritesReducer.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { ActionTypes } from 'enums/ActionTypes';
+import { Action, State } from './models';
+
+const favoritesReducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case ActionTypes.ADD:
+      return { favorites: [...state.favorites, action.payload] };
```
Comment (kamil-borecki-cc) [inline src/context/favorites/FavoritesReducer.ts:7]: **question** what about duplicates?

File: src/context/favorites/FavoritesProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+const FavoritesProvider = ({ children }: PropsWithChildren) => {
+  const [favorites, setFavorites] = useLocalStorage<string[]>('favorites');
+  const [state, dispatch] = useReducer(favoritesReducer, { favorites: favorites || [] });
+
+  const checkIfIsFavorite = (city: string): boolean => {
+    return state.favorites.includes(city);
+  };
+
+  useEffect(() => {
+    setFavorites(state.favorites);
```
Comment (kamil-borecki-cc) [inline src/context/favorites/FavoritesProvider.tsx:35]: **question** i dont see serialization array to string, how it is handled?

Comment (JakubLukaszCC) [reply src/context/favorites/FavoritesProvider.tsx:35]: Hook useLocalStorage handles JSON.stringify and JSON.parse before and after localStorage functions

File: src/context/favorites/FavoritesReducer.ts (hunk 3 above)
Comment (JakubLukaszCC) [reply src/context/favorites/FavoritesReducer.ts:7]: button makes add or remove action depending on `favorite` state, if city is favorite then it makes `remove` action, when city is not favorite it makes `add` action so i think it solves duplicates issue

File: src/context/favorites/FavoritesProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { useLocalStorage } from 'hooks/useLocalStorage';
+import { createContext, useContext, PropsWithChildren, useReducer, useEffect } from 'react';
+import { favoritesReducer } from './FavoritesReducer';
+import { State, Dispatch } from './models';
+
+const FavoriteContext = createContext<{ state: State; dispatch: Dispatch } | undefined>(undefined);
```
[the remark this answers is not in the document]
Comment (JakubLukaszCC) [reply src/context/favorites/FavoritesProvider.tsx]: we fixed this :)

File: src/components/FavoritesButton/FavoritesButton.test.tsx [hunk 6]
```diff
@@ -1,27 +1,38 @@
-import { render, screen } from 'test-utils';
+import { ADD } from 'context/favorites';
```
Comment (kamil-borecki-cc) [inline src/components/FavoritesButton/FavoritesButton.test.tsx]: **suggestion** test adding to favorites action here

**Review by klaudia-blazyczek-cc — APPROVED** [review]

File: src/context/favorites/Actions.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { ActionTypes } from 'enums/ActionTypes';
+import { Action } from './models';
+
+const ADD = (payload: string): Action => ({ type: ActionTypes.ADD, payload });
```
Comment (bednarskimichal) [inline src/context/favorites/Actions.ts]:
```suggestion
const add = (payload: string): Action => ({ type: ActionTypes.ADD, payload });
```
cc: @dkaczmarek-cc @kamil-borecki-cc, please let me know if i'm wrong, but I would like to suggest that method should be defined by uppercase letters here.

**Review by darya-karpovich-cc — APPROVED** [review]

File: src/context/favorites/Actions.ts (hunk 7 above)
Comment (adrian-konopnicki-CC) [reply src/context/favorites/Actions.ts]: We changed that :+1:

**Review by kamil-borecki-cc — APPROVED** [review]

**Pull request merged**